File: admin/src/lib/api/content.ts
/**
 * Typed API functions for all content CRUD endpoints.
 * All return { data, meta } from APIResponse.
 */
import { apiClient } from "./client";

export type PaginationMeta = {
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
};

export type APIListResponse<T> = {
  data: T[];
  meta: PaginationMeta;
};

export type APIItemResponse<T> = {
  data: T;
  meta?: null;
};

// ── Generic list/get/create/update/delete ─────────────────────────────────────
function makeContentApi<T, CreatePayload, UpdatePayload>(base: string) {
  return {
    list: (token: string, params?: Record<string, string | number>) => {
      const qs = params ? "?" + new URLSearchParams(params as Record<string, string>).toString() : "";
      return apiClient.get<APIListResponse<T>>(`${base}${qs}`, token);
    },
    get: (token: string, id: string) =>
      apiClient.get<APIItemResponse<T>>(`${base}/${id}`, token),
    create: (token: string, payload: CreatePayload) =>
      apiClient.post<APIItemResponse<T>>(base, payload, token),
    update: (token: string, id: string, payload: UpdatePayload) =>
      apiClient.patch<APIItemResponse<T>>(`${base}/${id}`, payload, token),
    delete: (token: string, id: string) =>
      apiClient.del<void>(`${base}/${id}`, token),
  };
}

// ── Type definitions mirroring Python schemas ─────────────────────────────────

export type ProductCategory = {
  id: string;
  category_name: string;
  slug: string;
  description: string | null;
  image_url: string | null;
  og_image_url: string | null;
  parent_id: string | null;
  sort_order: number;
  seo_title: string | null;
  seo_description: string | null;
  status: string;
  locale: string;
  created_at: string;
  updated_at: string;
};

export type Product = {
  id: string;
  product_name: string;
  slug: string;
  model_number: string;
  short_description: string;
  full_description: string | null;
  specifications: string | null;
  category_id: string;
  seo_title: string | null;
  seo_description: string | null;
  image_url: string | null;
  og_image_url: string | null;
  image_alt: string | null;
  status: string;
  locale: string;
  is_featured: boolean;
  display_priority: number;
  created_at: string;
  updated_at: string;
  published_at: string | null;
  gallery_images?: ProductGalleryImage[];
};

export type Application = {
  id: string;
  application_name: string;
  slug: string;
  industry: string;
  description: string | null;
  challenge: string | null;
  solution: string | null;
  hero_image_url: string | null;
  og_image_url: string | null;
  seo_title: string | null;
  seo_description: string | null;
  status: string;
  locale: string;
  sort_order: number;
  created_at: string;
  updated_at: string;
  published_at: string | null;
};

export type FAQItem = {
  id: string;
  variant_key?: string;
  question: string;
  answer: string;
  category_tag: string | null;
  locale: string;
  sort_order: number;
  status: string;
  created_at: string;
  updated_at: string;
};

export type ComparisonTopic = {
  id: string;
  topic_title: string;
  slug: string;
  summary: string | null;
  dimensions: string | null;
  conclusion: string | null;
  seo_title: string | null;
  seo_description: string | null;
  status: string;
  locale: string;
  sort_order: number;
  created_at: string;
  updated_at: string;
  published_at: string | null;
};

export type Certification = {
  id: string;
  cert_name: string;
  slug: string;
  issuer: string | null;
  cert_number: string | null;
  issued_at: string | null;
  expires_at: string | null;
  description: string | null;
  badge_image_url: string | null;
  document_url: string | null;
  locale: string;
  status: string;
  created_at: string;
  updated_at: string;
};

export type CTA = {
  id: string;
  cta_key: string;
  cta_type: string;
  headline: string;
  subheadline: string | null;
  button_label: string;
  button_action: string;
  button_url: string | null;
  bg_color: string | null;
  image_url: string | null;
  locale: string;
  status: string;
  sort_order: number;
  created_at: string;
  updated_at: string;
};

export type Page = {
  id: string;
  page_type: string;
  slug: string;
  title: string;
  subtitle: string | null;
  body: string | null;
  hero_image_url: string | null;
  seo_title: string | null;
  seo_description: string | null;
  og_image_url: string | null;
  canonical_url: string | null;
  structured_data: string | null;
  locale: string;
  status: string;
  noindex: boolean;
  entity_type: string | null;
  entity_id: string | null;
  created_at: string;
  updated_at: string;
  published_at: string | null;
};

export type RedirectRule = {
  id: string;
  from_path: string;
  to_path: string;
  status_code: 301 | 302;
  is_active: boolean;
  note: string;
  created_at: string;
  updated_at: string;
};

// ── API client instances ──────────────────────────────────────────────────────
const BASE = "/content";

export const categoriesApi = makeContentApi<ProductCategory, Partial<ProductCategory>, Partial<ProductCategory>>(`${BASE}/categories`);
export const productsApi = makeContentApi<Product, Partial<Product>, Partial<Product>>(`${BASE}/products`);
export const applicationsApi = makeContentApi<Application, Partial<Application>, Partial<Application>>(`${BASE}/applications`);
export const faqsApi = makeContentApi<FAQItem, Partial<FAQItem>, Partial<FAQItem>>(`${BASE}/faqs`);
export const comparisonsApi = makeContentApi<ComparisonTopic, Partial<ComparisonTopic>, Partial<ComparisonTopic>>(`${BASE}/comparisons`);
export const certificationsApi = makeContentApi<Certification, Partial<Certification>, Partial<Certification>>(`${BASE}/certifications`);

export type Capability = {
  id: string;
  capability_name: string;
  slug: string;
  icon_url: string | null;
  image_url: string | null;
  short_description: string;
  detail: string | null;
  metrics: string | null;
  category_tag: string | null;
  sort_order: number;
  locale: string;
  status: string;
  created_at: string;
  updated_at: string;
};

export const capabilitiesApi = makeContentApi<Capability, Partial<Capability>, Partial<Capability>>(`${BASE}/capabilities`);

export const ctasApi = makeContentApi<CTA, Partial<CTA>, Partial<CTA>>(`${BASE}/ctas`);
export const pagesApi = makeContentApi<Page, Partial<Page>, Partial<Page>>(`${BASE}/pages`);
export const redirectsApi = {
  list: (token: string, activeOnly = false) =>
    apiClient.get<RedirectRule[]>(`${BASE}/redirects?active_only=${String(activeOnly)}`, token),
  create: (token: string, payload: Partial<RedirectRule>) =>
    apiClient.post<RedirectRule>(`${BASE}/redirects`, payload, token),
  update: (token: string, id: string, payload: Partial<RedirectRule>) =>
    apiClient.patch<RedirectRule>(`${BASE}/redirects/${id}`, payload, token),
  delete: (token: string, id: string) =>
    apiClient.del<void>(`${BASE}/redirects/${id}`, token),
};
// ── Preview Token (1a.6.4) ────────────────────────────────────────────────────
export const previewApi = {
  createToken: (token: string, pageId: string) =>
    apiClient.post<{ token: string; expires_in_seconds: number; preview_url: string }>(
      `${BASE}/pages/${pageId}/preview-token`,
      {},
      token
    ),
};

// ── Publish / Unpublish ───────────────────────────────────────────────────────
export const localeDraftApi = {
  create: (token: string, entity: string, id: string, targetLocale?: string) =>
    apiClient.post<{
      action: string;
      entity: string;
      source_id: string;
      target_id: string;
      source_locale: string;
      target_locale: string;
      status: string;
    }>(`${BASE}/${entity}/${id}/locale-draft`, { target_locale: targetLocale ?? null }, token),
  createBatch: (token: string, entity: string, sourceIds: string[], targetLocale: string) =>
    apiClient.post<{
      entity: string;
      target_locale: string;
      requested: number;
      created_or_updated: number;
      failed: number;
      failures: Array<{ source_id: string; status_code: number; message: string }>;
      published: number;
    }>(`${BASE}/locale-drafts/batch`, {
      entity,
      source_ids: sourceIds,
      target_locale: targetLocale,
    }, token),
};

export const localeCoverageApi = {
  get: (token: string, targetLocale?: string) =>
    apiClient.get<{
      source_locale: string;
      target_locale: string;
      source_total: number;
      translated: number;
      overall_coverage_pct: number | null;
      missing: number;
      draft: number;
      stale: number;
      unpaired: number;
      entities: Array<{ entity: string; missing_keys: string[]; missing_ids: string[]; missing_count: number; draft: number; stale: number; unpaired: number; unpaired_keys: string[]; source_total: number; translated: number; published: number; coverage_pct: number | null; published_pct: number | null }>;
      policy: string;
    }>(`${BASE}/locale-coverage${targetLocale ? `?target_locale=${encodeURIComponent(targetLocale)}` : ""}`, token),
  settings: (token: string) =>
    apiClient.get<{
      source_locale: string;
      content_locales: Array<{ content_locale: string; route_locale: string; label: string; native_label: string; public_shell_ready: boolean }>;
      public_shell_locales: string[];
      policy: string;
    }>(`${BASE}/locale-settings`, token),
};

export const publishApi = {
  publish: (token: string, entity: string, id: string) =>
    apiClient.post<{ detail: string; id: string }>(`${BASE}/${entity}/${id}/publish`, {}, token),
  unpublish: (token: string, entity: string, id: string) =>
    apiClient.post<{ detail: string; id: string }>(`${BASE}/${entity}/${id}/unpublish`, {}, token),
};

// ── M2M Relations ─────────────────────────────────────────────────────────────
export const relationsApi = {
  listProductApplications: (token: string, productId: string) =>
    apiClient.get<{ id: string; name: string; slug: string }[]>(`${BASE}/products/${productId}/applications`, token),
  linkProductApplication: (token: string, productId: string, appId: string) =>
    apiClient.post<{ detail: string }>(`${BASE}/products/${productId}/applications/${appId}`, {}, token),
  unlinkProductApplication: (token: string, productId: string, appId: string) =>
    apiClient.del<void>(`${BASE}/products/${productId}/applications/${appId}`, token),

  listProductCertifications: (token: string, productId: string) =>
    apiClient.get<{ id: string; name: string; slug: string }[]>(`${BASE}/products/${productId}/certifications`, token),
  linkProductCertification: (token: string, productId: string, certId: string) =>
    apiClient.post<{ detail: string }>(`${BASE}/products/${productId}/certifications/${certId}`, {}, token),
  unlinkProductCertification: (token: string, productId: string, certId: string) =>
    apiClient.del<void>(`${BASE}/products/${productId}/certifications/${certId}`, token),

  listProductFAQs: (token: string, productId: string) =>
    apiClient.get<{ id: string; name: string; slug: string }[]>(`${BASE}/products/${productId}/faqs`, token),
  linkProductFAQ: (token: string, productId: string, faqId: string) =>
    apiClient.post<{ detail: string }>(`${BASE}/products/${productId}/faqs/${faqId}`, {}, token),
  unlinkProductFAQ: (token: string, productId: string, faqId: string) =>
    apiClient.del<void>(`${BASE}/products/${productId}/faqs/${faqId}`, token),
  listApplicationFAQs: (token: string, applicationId: string) =>
    apiClient.get<{ id: string; name: string; slug: string }[]>(`${BASE}/applications/${applicationId}/faqs`, token),
  linkApplicationFAQ: (token: string, applicationId: string, faqId: string) =>
    apiClient.post<{ detail: string }>(`${BASE}/applications/${applicationId}/faqs/${faqId}`, {}, token),
  unlinkApplicationFAQ: (token: string, applicationId: string, faqId: string) =>
    apiClient.del<void>(`${BASE}/applications/${applicationId}/faqs/${faqId}`, token),
};

// ── Content Asset ─────────────────────────────────────────────────────────────
export type ContentAsset = {
  id: string;
  original_filename: string;
  r2_key: string;
  public_url: string;
  mime_type: string;
  file_size_bytes: number;
  asset_type: "image" | "pdf" | "cad" | "document" | "other";
  alt_text: string | null;
  title: string | null;
  is_indexable: boolean;
  index_status?: "not_indexed" | "not_applicable" | "pending" | "indexed" | "needs_ocr" | "withdrawn";
  index_error?: string | null;
  seo_title: string | null;
  product_id: string | null;
  page_id: string | null;
  display_order: number;
  uploaded_by: string;
  created_at: string;
};

export type ProductGalleryImage = {
  id: string;
  public_url: string;
  alt_text: string | null;
  display_order: number;
};

export const assetsApi = {
  list: (token: string, params?: Record<string, string | number>) => {
    const qs = params ? "?" + new URLSearchParams(params as Record<string, string>).toString() : "";
    return apiClient.get<APIListResponse<ContentAsset>>(`${BASE}/assets${qs}`, token);
  },
  upload: (token: string, formData: FormData) =>
    apiClient.postForm<ContentAsset>(`${BASE}/assets`, formData, token),
  delete: (token: string, id: string) =>
    apiClient.del<void>(`${BASE}/assets/${id}`, token),
  update: (token: string, id: string, payload: Partial<Pick<ContentAsset, "alt_text" | "display_order" | "product_id" | "is_indexable" | "title">>) =>
    apiClient.patch<ContentAsset>(`${BASE}/assets/${id}`, payload, token),
  updateAlt: (token: string, id: string, altText: string) =>
    apiClient.patch<ContentAsset>(`${BASE}/assets/${id}`, { alt_text: altText }, token),
};

export const productGalleryApi = {
  reorder: (token: string, productId: string, items: Array<{ id: string; display_order: number }>) =>
    apiClient.put<APIItemResponse<Product>>(`${BASE}/products/${productId}/gallery`, { items }, token),
};

File: api/app/schemas/content.py
import uuid
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


class ApplicationCreate(BaseModel):
    application_name: str = Field(max_length=100)
    slug: str = Field(max_length=100, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    industry: str = Field(max_length=60)
    description: Optional[str] = None
    challenge: Optional[str] = None
    solution: Optional[str] = None
    hero_image_url: Optional[str] = Field(default=None, max_length=500)
    og_image_url: Optional[str] = Field(default=None, max_length=500)
    seo_title: Optional[str] = Field(default=None, max_length=70)
    seo_description: Optional[str] = Field(default=None, max_length=160)
    status: str = Field(default="draft")
    locale: str = Field(default="en", max_length=5)
    sort_order: int = Field(default=0, ge=0)


class ApplicationUpdate(BaseModel):
    application_name: Optional[str] = Field(default=None, max_length=100)
    slug: Optional[str] = Field(default=None, max_length=100, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    industry: Optional[str] = Field(default=None, max_length=60)
    description: Optional[str] = None
    challenge: Optional[str] = None
    solution: Optional[str] = None
    hero_image_url: Optional[str] = Field(default=None, max_length=500)
    og_image_url: Optional[str] = Field(default=None, max_length=500)
    seo_title: Optional[str] = Field(default=None, max_length=70)
    seo_description: Optional[str] = Field(default=None, max_length=160)
    status: Optional[str] = None
    locale: Optional[str] = Field(default=None, max_length=5)
    sort_order: Optional[int] = Field(default=None, ge=0)


class ApplicationRead(BaseModel):
    id: uuid.UUID
    application_name: str
    slug: str
    industry: str
    description: Optional[str]
    challenge: Optional[str]
    solution: Optional[str]
    hero_image_url: Optional[str]
    og_image_url: Optional[str]
    seo_title: Optional[str]
    seo_description: Optional[str]
    status: str
    locale: str
    sort_order: int
    created_at: datetime
    updated_at: datetime
    published_at: Optional[datetime]

    model_config = {"from_attributes": True}


# ── FAQItem schemas ───────────────────────────────────────────────────────────

class FAQItemCreate(BaseModel):
    question: str = Field(max_length=300)
    answer: str
    category_tag: Optional[str] = Field(default=None, max_length=60)
    locale: str = Field(default="en", max_length=5)
    sort_order: int = Field(default=0, ge=0)
    status: str = Field(default="draft")
    variant_key: Optional[str] = Field(default=None, max_length=80)


class FAQItemUpdate(BaseModel):
    question: Optional[str] = Field(default=None, max_length=300)
    answer: Optional[str] = None
    category_tag: Optional[str] = Field(default=None, max_length=60)
    locale: Optional[str] = Field(default=None, max_length=5)
    sort_order: Optional[int] = Field(default=None, ge=0)
    status: Optional[str] = None
    variant_key: Optional[str] = Field(default=None, max_length=80)


class FAQItemRead(BaseModel):
    id: uuid.UUID
    variant_key: str
    question: str
    answer: str
    category_tag: Optional[str]
    locale: str
    sort_order: int
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── ComparisonTopic schemas ───────────────────────────────────────────────────

class ComparisonTopicCreate(BaseModel):
    topic_title: str = Field(max_length=120)
    slug: str = Field(max_length=120, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    summary: Optional[str] = Field(default=None, max_length=500)
    dimensions: Optional[str] = None   # JSON
    conclusion: Optional[str] = None
    seo_title: Optional[str] = Field(default=None, max_length=70)
    seo_description: Optional[str] = Field(default=None, max_length=160)
    status: str = Field(default="draft")
    locale: str = Field(default="en", max_length=5)
    sort_order: int = Field(default=0, ge=0)


class ComparisonTopicUpdate(BaseModel):
    topic_title: Optional[str] = Field(default=None, max_length=120)
    slug: Optional[str] = Field(default=None, max_length=120, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    summary: Optional[str] = Field(default=None, max_length=500)
    dimensions: Optional[str] = None
    conclusion: Optional[str] = None
    seo_title: Optional[str] = Field(default=None, max_length=70)
    seo_description: Optional[str] = Field(default=None, max_length=160)
    status: Optional[str] = None
    locale: Optional[str] = Field(default=None, max_length=5)
    sort_order: Optional[int] = Field(default=None, ge=0)


class ComparisonTopicRead(BaseModel):
    id: uuid.UUID
    topic_title: str
    slug: str
    summary: Optional[str]
    dimensions: Optional[str]
    conclusion: Optional[str]
    seo_title: Optional[str]
    seo_description: Optional[str]
    status: str
    locale: str
    sort_order: int
    created_at: datetime
    updated_at: datetime
    published_at: Optional[datetime]

    model_config = {"from_attributes": True}


# ── Certification schemas ─────────────────────────────────────────────────────

class CertificationCreate(BaseModel):
    cert_name: str = Field(max_length=100)
    slug: Optional[str] = Field(default=None, max_length=120, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    issuer: Optional[str] = Field(default=None, max_length=120)
    cert_number: Optional[str] = Field(default=None, max_length=80)
    issued_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    description: Optional[str] = None
    badge_image_url: Optional[str] = Field(default=None, max_length=500)
    document_url: Optional[str] = Field(default=None, max_length=500)
    locale: str = Field(default="en", max_length=5)
    status: str = Field(default="active")


class CertificationUpdate(BaseModel):
    cert_name: Optional[str] = Field(default=None, max_length=100)
    slug: Optional[str] = Field(default=None, max_length=120, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    issuer: Optional[str] = Field(default=None, max_length=120)
    cert_number: Optional[str] = Field(default=None, max_length=80)
    issued_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    description: Optional[str] = None
    badge_image_url: Optional[str] = Field(default=None, max_length=500)
    document_url: Optional[str] = Field(default=None, max_length=500)
    locale: Optional[str] = Field(default=None, max_length=5)
    status: Optional[str] = None


class CertificationRead(BaseModel):
    id: uuid.UUID
    cert_name: str
    slug: str
    issuer: Optional[str]
    cert_number: Optional[str]
    issued_at: Optional[datetime]
    expires_at: Optional[datetime]
    description: Optional[str]
    badge_image_url: Optional[str]
    document_url: Optional[str]
    locale: str
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Capability schemas ────────────────────────────────────────────────────────

class CapabilityCreate(BaseModel):
    capability_name: str = Field(max_length=100)
    slug: str = Field(max_length=100, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    icon_url: Optional[str] = Field(default=None, max_length=500)
    image_url: Optional[str] = Field(default=None, max_length=500)
    short_description: str = Field(max_length=200)
    detail: Optional[str] = None
    metrics: Optional[str] = None   # JSON
    category_tag: Optional[str] = Field(default=None, max_length=60)
    sort_order: int = Field(default=0, ge=0)
    locale: str = Field(default="en", max_length=5)
    status: str = Field(default="draft")


class CapabilityUpdate(BaseModel):
    capability_name: Optional[str] = Field(default=None, max_length=100)
    slug: Optional[str] = Field(default=None, max_length=100, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    icon_url: Optional[str] = Field(default=None, max_length=500)
    image_url: Optional[str] = Field(default=None, max_length=500)
    short_description: Optional[str] = Field(default=None, max_length=200)
    detail: Optional[str] = None
    metrics: Optional[str] = None
    category_tag: Optional[str] = Field(default=None, max_length=60)
    sort_order: Optional[int] = Field(default=None, ge=0)
    locale: Optional[str] = Field(default=None, max_length=5)
    status: Optional[str] = None


class CapabilityRead(BaseModel):
    id: uuid.UUID
    capability_name: str
    slug: str
    icon_url: Optional[str]
    image_url: Optional[str]
    short_description: str
    detail: Optional[str]
    metrics: Optional[str]
    category_tag: Optional[str]
    sort_order: int
    locale: str
    status: str
    created_at: datetime
    updated_at: datetime
    published_at: Optional[datetime]

    model_config = {"from_attributes": True}


# ── CTA schemas ───────────────────────────────────────────────────────────────

class CTACreate(BaseModel):
    cta_key: str = Field(max_length=60)
    cta_type: str = Field(max_length=30)
    headline: str = Field(max_length=120)
    subheadline: Optional[str] = Field(default=None, max_length=240)
    button_label: str = Field(max_length=60)
    button_action: str = Field(max_length=30)
    button_url: Optional[str] = Field(default=None, max_length=500)
    bg_color: Optional[str] = Field(default=None, max_length=20)
    image_url: Optional[str] = Field(default=None, max_length=500)
    locale: str = Field(default="en", max_length=5)
    status: Literal["draft", "published", "archived"] = "draft"
    sort_order: int = Field(default=0, ge=0)


class CTAUpdate(BaseModel):
    cta_key: Optional[str] = Field(default=None, max_length=60)
    cta_type: Optional[str] = Field(default=None, max_length=30)
    headline: Optional[str] = Field(default=None, max_length=120)
    subheadline: Optional[str] = Field(default=None, max_length=240)
    button_label: Optional[str] = Field(default=None, max_length=60)
    button_action: Optional[str] = Field(default=None, max_length=30)
    button_url: Optional[str] = Field(default=None, max_length=500)
    bg_color: Optional[str] = Field(default=None, max_length=20)
    image_url: Optional[str] = Field(default=None, max_length=500)
    locale: Optional[str] = Field(default=None, max_length=5)
    status: Optional[Literal["draft", "published", "archived"]] = None
    sort_order: Optional[int] = Field(default=None, ge=0)


class CTARead(BaseModel):
    id: uuid.UUID
    cta_key: str
    cta_type: str
    headline: str
    subheadline: Optional[str]
    button_label: str
    button_action: str
    button_url: Optional[str]
    bg_color: Optional[str]
    image_url: Optional[str]
    locale: str
    status: str
    sort_order: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Page schemas ──────────────────────────────────────────────────────────────

class PageCreate(BaseModel):
    page_type: str = Field(max_length=40)
    slug: str = Field(max_length=120, pattern=r"^[a-z0-9]+(?:[/-][a-z0-9]+)*$")
    title: str = Field(max_length=120)
    subtitle: Optional[str] = Field(default=None, max_length=240)
    body: Optional[str] = None
    hero_image_url: Optional[str] = Field(default=None, max_length=500)
    seo_title: Optional[str] = Field(default=None, max_length=70)
    seo_description: Optional[str] = Field(default=None, max_length=160)
    og_image_url: Optional[str] = Field(default=None, max_length=500)
    canonical_url: Optional[str] = Field(default=None, max_length=500)
    structured_data: Optional[str] = None
    locale: str = Field(default="en", max_length=5)
    status: str = Field(default="draft")
    noindex: bool = False
    entity_type: Optional[str] = Field(default=None, max_length=40)
    entity_id: Optional[uuid.UUID] = None


class PageUpdate(BaseModel):
    page_type: Optional[str] = Field(default=None, max_length=40)
    slug: Optional[str] = Field(default=None, max_length=120)
    title: Optional[str] = Field(default=None, max_length=120)
    subtitle: Optional[str] = Field(default=None, max_length=240)
    body: Optional[str] = None
    hero_image_url: Optional[str] = Field(default=None, max_length=500)
    seo_title: Optional[str] = Field(default=None, max_length=70)
    seo_description: Optional[str] = Field(default=None, max_length=160)
    og_image_url: Optional[str] = Field(default=None, max_length=500)
    canonical_url: Optional[str] = Field(default=None, max_length=500)
    structured_data: Optional[str] = None
    locale: Optional[str] = Field(default=None, max_length=5)
    status: Optional[str] = None
    noindex: Optional[bool] = None
    entity_type: Optional[str] = Field(default=None, max_length=40)
    entity_id: Optional[uuid.UUID] = None


class PageRead(BaseModel):
    id: uuid.UUID
    page_type: str
    slug: str
    title: str
    subtitle: Optional[str]
    body: Optional[str]
    hero_image_url: Optional[str]
    seo_title: Optional[str]
    seo_description: Optional[str]
    og_image_url: Optional[str]
    canonical_url: Optional[str]
    structured_data: Optional[str]
    locale: str
    status: str
    noindex: bool
    entity_type: Optional[str]
    entity_id: Optional[uuid.UUID]
    created_at: datetime
    updated_at: datetime
    published_at: Optional[datetime]

    model_config = {"from_attributes": True}

